x-pack/plugins/security/server/authorization/authorization_service.ts

```diff
@@ -0,0 +1,218 @@
+    getSpacesService,
+  }: AuthorizationServiceSetupParams): AuthorizationServiceSetup {
+    this.logger = loggers.get('authorization');
+    this.license = license;
+    this.status = status;
+    this.applicationName = `${APPLICATION_PREFIX}${kibanaIndexName}`;
+
+    const mode = authorizationModeFactory(license);
+    const actions = new Actions(packageVersion);
+    this.privileges = privilegesFactory(actions, features, license);
+
+    const checkPrivilegesWithRequest = checkPrivilegesWithRequestFactory(
+      actions,
+      clusterClient,
+      this.applicationName
+    );
+
+    const authz = {
+      actions,
+      applicationName: this.applicationName,
+      mode,
+      privileges: this.privileges,
+      checkPrivilegesWithRequest,
+      checkPrivilegesDynamicallyWithRequest: checkPrivilegesDynamicallyWithRequestFactory(
+        checkPrivilegesWithRequest,
+        getSpacesService
+      ),
+      checkSavedObjectsPrivilegesWithRequest: checkSavedObjectsPrivilegesWithRequestFactory(
+        checkPrivilegesWithRequest,
+        getSpacesService
+      ),
+    };
+
+    capabilities.registerSwitcher(
+      async (request: KibanaRequest, uiCapabilities: UICapabilities) => {
+        // If we have a license which doesn't enable security, or we're a legacy user we shouldn't
+        // disable any ui capabilities
+        if (!mode.useRbacForRequest(request)) {
+          return uiCapabilities;
+        }
+
+        const disableUICapabilities = disableUICapabilitiesFactory(
+          request,
+          features.getFeatures(),
+          this.logger,
+          authz
+        );
+
+        if (!request.auth.isAuthenticated) {
+          return disableUICapabilities.all(uiCapabilities);
+        }
+
+        return await disableUICapabilities.usingPrivileges(uiCapabilities);
+      }
+    );
+
+    initAPIAuthorization(http, authz, loggers.get('api-authorization'));
+    initAppAuthorization(http, authz, loggers.get('app-authorization'), features);
+
+    return authz;
+  }
+
+  start({ clusterClient, features }: AuthorizationServiceStartParams) {
+    const allFeatures = features.getFeatures();
+    validateFeaturePrivileges(allFeatures);
+    validateReservedPrivileges(allFeatures);
+
+    this.registerPrivileges(clusterClient);
+  }
+
+  stop() {
+    if (this.statusSubscription !== undefined) {
+      this.statusSubscription.unsubscribe();
+      this.statusSubscription = undefined;
+    }
+  }
+
+  private registerPrivileges(clusterClient: IClusterClient) {
+    const RETRY_SCALE_DURATION = 100;
+    const RETRY_TIMEOUT_MAX = 10000;
+    const retries$ = new BehaviorSubject(0);
+    let retryTimeout: NodeJS.Timeout;
+
+    // Register cluster privileges once Elasticsearch is available and Security plugin is enabled.
+    this.statusSubscription = combineLatest([
+      this.status.core$,
+      this.license.features$,
+      retries$.asObservable(),
+    ])
+      .pipe(
+        takeWhile(() => !retries$.isStopped),
```

Comment:
I don't think this exhibits the behavior we want in all cases. As soon as privileges are successfully registered, we complete the `retries$` observable, which prevents registration from ever happening again, even if there is a change to core's status or the license status.

A couple examples:

### Restarting dev cluster

1) Startup Kibana and ES, and wait for privilege registration to complete
2) Stop ES, leave Kibana running
3) Start up a *fresh* ES instance

We would expect Kibana to re-register application privileges in this case. Without registration, only the superuser will be able to authenticate to Kibana


### Upgrading from Basic to Gold+ license
After upgrading (or downgrading), we need to reattempt privilege registration because the privileges we register are dependent on the current license level. Upgrading from basic to gold+ means that we need to register all sub-feature privileges, and downgrading to basic means that we need to de-register all sub-feature privileges.

Reply:
Sorry for going back and forth on this one and thanks for laying out this for me! I didn't realize it works like this currently while reading the code. Especially the case with restarting cluster as this event can be potentially missed by Kibana depending on circumstances and `elasticsearch.healthCheck.delay`.

Let me dig and give it another try.

Reply:
I _believe_ it now covers all discussed use cases + I added a API integration test for a license downgrade scenario. Would you mind taking a look whenever you have time @legrego ?